### PR TITLE
Session 06

### PR DESCRIPTION
**Pull request by rkenyon3**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,7 +5,7 @@ edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
-[dependencies]
+[dependencies]  # daniel suggested these would be better in dev-dependencies but if I try it complains
```
Comment (kinnison) [inline Cargo.toml]: Not sure what this comment is about?

File: bft_interp/src/lib.rs [hunk 2]
```diff
@@ -142,58 +147,96 @@ where
...
                 Err(VMError::from((bad_instruction, error)))
             }
         }
     }
+
+    /// Get the next program instruction index based on the value of the cell under the head.
+    /// If the cell is zero, return the index of the instruction after the matching ].
+    /// If the cell is not zero, return the index of the next instruction after this one.
+    fn conditional_jump_forward(&self) -> Result<usize, VMError> {
+        match self.program.jump_map()[self.program_counter] {
```
Comment (kinnison) [inline bft_interp/src/lib.rs]: I think I'd just make the jump map be infallible, it's not like you'll bother looking stuff up in places which aren't jumps.

File: bft_types/src/lib.rs [hunk 3]
```diff
@@ -199,10 +199,15 @@ impl BfProgram {
     }
 
     /// The instructions that make up this program
-    pub fn instructions(&self) -> &[LocalisedInstruction] {
+    pub fn localised_instructions(&self) -> &[LocalisedInstruction] {
         &self.instructions
     }
 
+    /// The vector of indexes '[' and ']' may jump to
+    pub fn jump_map(&self) -> &Vec<Option<usize>> {
```
Comment (kinnison) [inline bft_types/src/lib.rs]: I don't like this function signature, I'd suggest:

`pub fn jump_target(&self, ip: usize) -> usize`

File: bft_types/src/lib.rs [hunk 4]
```diff
@@ -220,23 +225,28 @@ impl BfProgram {
     ///# }
     /// ```
     pub fn analyse_program(&mut self) -> Result<(), String> {
+        // Ensure we don't double the jump map if analyse gets called twice.
+        // Might be better as a flag in the struct?
```
Comment (kinnison) [inline bft_types/src/lib.rs]: Make this non-public and call it from your initialisation methods and this wouldn't happen.

File: bft_types/src/lib.rs [hunk 5]
```diff
@@ -264,6 +274,7 @@ impl BfProgram {
     }
 }
 
+// TODO: find out why cargo test --all from ../../.. isn't runnning these tests :(
```
Comment (kinnison) [inline bft_types/src/lib.rs]: What's the error?

File: src/main.rs [hunk 6]
```diff
@@ -10,14 +10,51 @@
...
+        }
+    }
+}
+
+impl<T: Write> Write for WriterWithTrailingNewline<T> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        if let Some(last_byte) = buf.last() {
+            self.last_byte = *last_byte;
+        }
+        self.inner_writer.write(buf)
```
Comment (kinnison) [inline src/main.rs]: This isn't going to guarantee to write everything - either use write_all or else use write, but then get the nbytes from that to decide which to check for newlineness in your buffer.  See me if you can't work out how to do these.

File: src/main.rs [hunk 7]
```diff
@@ -10,14 +10,51 @@
...
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.inner_writer.flush()
+    }
+}
+
+impl<T: Write> Drop for WriterWithTrailingNewline<T> {
+    fn drop(&mut self) {
+        if self.last_byte != b'\n' {
+            let buf = [b'\n'];
+            let _ = self.inner_writer.write(&buf);
```
Comment (kinnison) [inline src/main.rs]:
```suggestion
            writeln!(self.inner_writer).expect("Unable to write newline");
```

File: bft_types/src/lib.rs (hunk 5 above)
Comment (rkenyon3) [reply bft_types/src/lib.rs]: Whoops, forgot to remove that comment. It turned out it was stopping on a broken doctest elsewhere. Once that was fixed it picked up the rest

**Pull request merged**